src/wolff_producer.erl

```diff
@@ -376,13 +383,20 @@ send_to_kafka(#{sent_reqs := Sent,
     replayq:pop(Q, #{bytes_limit => BytesLimit, count_limit => 999999999}),
   {FlatBatch, Calls} = get_flat_batch(Items, [], []),
   [_ | _] = FlatBatch, %% assert
+  BatchSize = erlang:length(FlatBatch),
+  wolff_metrics:queuing_change(Config, -BatchSize),
+  wolff_metrics:inflight_change(Config, BatchSize),
   Req = kpro_req_lib:produce(Vsn, Topic, Partition, FlatBatch,
                              #{ack_timeout => AckTimeout,
                                required_acks => RequiredAcks,
                                compression => Compression
                               }),
   St1 = St0#{replayq := NewQ, ?linger_expire_timer := false},
-  NewSent = ?sent_req(Req, QAckRef, Calls),
+  NewSent = #{request => Req,
+              q_ack_ref => QAckRef,
+              calls => Calls,
+              batch_size => BatchSize,
+              attempts => 1},
   St2 = St1#{sent_reqs := queue:in(NewSent, Sent),
              sent_reqs_count := SentCount + 1
```

Comment:
if we keep the total number of accumulated `FlatBatche`s , we can report absolute value here.

Reply:
Ok, so we keep total number of `queuing` and `inflight` messages internally and report these numbers when changed instead of reporting the deltas...

Reply:
yes, otherwise the handler would have to keep track of ALL events wihtout missing a single one to get the truth.

couters are fine to only report delta, since a more common usage pattern of counters is to see how much incremented in a time window, as long as there is no missing events in the given window, it's the truth.

Reply:
The current counter's API for EMQX (see `emqx_metrics_worker`) don't have a function for setting a counter/gauge to a specific value. Only to change based on delta. I think there is a performance reason for this as we use the Erlang's counters API that internally use decentralized counters (https://www.erlang.org/doc/man/counters.html see write_concurrency and also the note for the put function). Reading the counter value is quite expensive and it may not be exact (as it traverses all internal counters without blocking updates). Similarly setting the counter to a specific value also requires traversal of all "parts" of the counter. For the wolff events there might not be a good reason to use "write_concurrency" (i.e., decentralized) counters (at least with the current implementation) as all events for a particular counter is triggered by a single process but I think there is a case to be made for events to report deltas instead of the total values as this is more future proof (in case we want to do some parallel processing in the wolff producer in the feature), and it does not force us to make the emqx counters API more complex. What do you think?

Reply:
After the discussion: this formular only tells if enter is growing faster or slower than exit.

Reply:
I and @zmstone have discussed this. Missing counter updates because a producer process is killed should not be a problem for EMQX. If one process is killed, all other processes should be killed and all gauges (and maybe also all counters) belonging to a resource should be reset to zero. If an event is missing for some other reason, it is probably a bug that should be fixed.